Contributing to open source blog post (#186)
ombulabs/blog: _posts/2019-06-3-4-contribute-open-source.markdown

---
layout: post

> title: "Contributing to open source"

Reviewer: @bronzdoc Try titleizing this title. Follow the capitalization we have been using in past blog articles.

date: 2019-06-03 09:00:00
categories: ["open-source"]
author: "bronzdoc"
---

> Contributing to open source might be scary, you might be thinking that you PR will not have a good quality or that people will judge you by your code, or that fixing that little typo is not worth it. I had all this questions in my head before i did my first contribution to open source and stoped me for a long time to contribute.

Reviewer: > Contributing to open source might be scary, you might be thinking that you PR will not have a good quality or that people will judge you by your code, or that fixing that little typo is not worth it.  

Might sound better like: 

> Contributing to open source might be scary, you might think that you pull request (PR) is not good enough; that people will judge you by your code; or that fixing that little typo is not worth it.

Reviewer: Also: 

> I had all this questions in my head before i did my first contribution to open source and stoped me for a long time to contribute.

Might sound better like this: 

> I had all these questions in my head before I submitted my first contribution to an open source project and it stopped me many times.

> So my suggestions is, start small, look for low hanging issues in a project you want to contribute, start looking at the issues and see if they have been solved previously and suggest a fix, by doing this you will start learning the code base and it will give you confidence to start tackling more complex issues.

Reviewer: @bronzdoc Punctuations could be a little better like this: 

> So my suggestion is: start small. Look for low hanging issues in a project you want to contribute to; start reviewing issues; and see if they have been solved in the past. Then you can suggest a fix. If you take it one small issue at a time, you will start learning the code base and it will give you the confidence you need to start tackling more complex issues.

This is a set of steps that would help you get started in your open source journey:

### If you think you found a bug:

> 1. Search in Stack Overflow

Reviewer: @bronzdoc I would start in Google. Sometimes there is a blog post about it which is not in SO or GitHub Issues.

Author: I modified that  line to `look in the web` i don't want to be specific about a source of looking for things.

2. Search in GitHub's Issues

> 3. If I don't find anything anywhere, submit an issue

Reviewer: @bronzdoc Here you are switching between "you" and "I" so I would pick one and stick with it. I would stick with "you"

Reviewer: I would try to be as explicit as possible in the issue submission form.

> 4. If I know a patch that works, submit a PR

Reviewer: @bronzdoc I'd say, monkey patch it in your project first, make sure it works, submit a PR.

Author: That might be implied, i don't want to go into detail about what is "that works"

> 5. Wait for feedback

Reviewer: @bronzdoc Sometimes you don't know enough to submit a PR or to patch the library, so I think it is valid to add more info to an existing issue (if there is one)

### If you think something will be a nice feature:

> 1. Search in GitHub's Issues if that feature is already proposed

Reviewer: If it is, but you think it could be tweaked, I'd add a comment to mention how I would want it to work.

2. If you don't find anything in the issue tracker, submit an issue explaining why you think your feature will add value to the project
3. Wait for feeback
4. if you got a positive feedback about your feature, submit a PR
5. Wait for feddback

> These steps are not written in gold, but i think they might help you get you started.

Reviewer: @bronzdoc I think you can say these are the steps that we follow at Ombu Labs.

Author: @etagwerker something like the following? 
`These steps are not written in gold, but it is what we use at Ombulabs and we think they might help you get you started.`

> There are situations where i personally wouldn't submit an issue first, i.e typos, formatting issues, etc

Reviewer: @bronzdoc Be careful when using 'i' and try to always use "I"

---
layout: post
title: "Contributing To Open Source"
date: 2019-06-03 09:00:00
categories: ["open-source"]
author: "bronzdoc"
---

Contributing to open source might be scary, you might think that you pull request (PR) is not good enough; that people will judge you by your code; or that fixing that little typo is not worth it.

I had all these questions in my head before I submitted my first contribution to an open source project and it stopped me many times.

> So my suggestion is: start small. Look for low hanging issues in a project you want to contribute to; start reviewing issues; and see if they have been solved in the past. Then you can suggest a fix. If you take it one small issue at a time, you will start learning the code base and it will give you the confidence you need to start tackling more complex issues.

Author: That would be awesome, i didn't found any reference to that feature, could you paste the link? t

> So my suggestion is: start small. Look for low hanging issues in a project you want to contribute to; start reviewing issues; and see if they have been solved in the past. Then you can suggest a fix. If you take it one small issue at a time, you will start learning the code base and it will give you the confidence you need to start tackling more complex issues.

Author: I couldn't find the link of what you mentioned @etagwerker but i added a `codetriage` link

Before getting started make sure you know the basics of version control and [git](https://git-scm.com/book/en/v1/Getting-Started-Git-Basics)

> Note: I'll be focusing on GitHub as the platform to contribute since is the most popular.

## This is a set of steps that would help you get started in your open source journey:

### If you think you found a bug

> 1. Look in the web to check the bug aleady exists

Author: What if the use Firefox 🔥 🦊 ?  😝

> 1. Look in the web to check the bug aleady exists

Reviewer: Even Firefox users use google.com though, right? 😉

> 1. Look in the web to check the bug aleady exists

Author: Thats right! 😂 i was directly associating google.com with chrome 🤦‍♂

2. If you don't find anything anywhere, submit an issue to the project repository explaining what you found and how you expected the program to behave and what you are getting

> 3. If you know a patch that works, submit a PR

Author: Implementation details could be discussed in the PR and is the feedback loop we mention trough the process of contributing.

> 3. If you know a patch that works, submit a PR

Author: ah i missread this! will add it!

4. Wait for feedback and make changes to you PR if the maintainers of the project have any until the PR is ready to merge!

### If you think something will be a nice feature

1. Search in Google if the feaure was already proposed or what other people think of your feature; if the feature is already proposed and you think you can improve it, comment your ideas about it.
2. If you don't find anything in the web, issue tracker, etc, submit an issue explaining why you think your feature will add value to the project

> 3. Wait for feeback from the project maintainers and learn about why your feature is or not addecuate to the project

Reviewer 2: Small typos:
```suggestion
3. Wait for feedback from the project maintainers and learn about why your feature is or not adequate to the project
```

4. If you got a positive feedback about your feature, start working on your feature and dont' get scared to ask for help if you get stuck. Finaly submit a PR!
5. Wait for feddback again! you are getting closer! maintainers of the project will give you more feedback about the feature; don't take the feedback personal and discuss the feedback if you disagree with something it might be a good way to learn from one and other.

### Triaging issue

> You might or not have a feature of issue to fix, another good way you could start contributing is by simply looking at the github issues, start checking if there are issues that are solved and notify the maintainers that the issue should be closed or if you see a feature maybe you could give feedback about it or maybe you could start reviewing existing PR's! the goal here is to lose the fear and to motivate you get involve in the project you choosed.

Reviewer: ```suggestion
You might not have a feature idea or issue to report. That shouldn't stop you from contributing to an open source project. You could triage open issues and pull requests. That's another great way you could start contributing. Start checking if there are issues that are already solved and notify maintainers that the issue should be closed.  You could give feedback about a feature proposal. You could start reviewing existing PRs! The main goal here is to lose your fear and motivate yourself to get involved in your favorite project.
```

## Projects to contribute

> You can find a list of "Awesome first PR Opportunities" [here](https://github.com/MunGell/awesome-for-beginners)

Reviewer: @bronzdoc Maybe mention CodeTriage?

## Conclusion

> These steps are not written in gold, but I think they might help you get you started.

Reviewer: ```suggestion
These steps are not written in gold, but I think they might help you get started.
```

> There are situations where I personally wouldn't submit an issue first, i.e typos, formatting issues, etc

Reviewer: ```suggestion
There a few cases where I wouldn't submit an issue first (i.e typos, formatting issues) 
```

> but this depends on how large is the change and how well you know the codebase.

Reviewer: ```suggestion
but this depends on the complexity of the change and how well you know the codebase.
```

Before getting started make sure you know the basics of version control and [git](https://git-scm.com/book/en/v1/Getting-Started-Git-Basics)

> Note: I'll be focusing on GitHub as the platform to contribute considering it's the most popular social coding community.

## This is a set of steps that would help you get started in your open source journey:

### If you think you found a bug

1. Search for issues in Google and in GitHub Issues to determine whether the bug already exists or it is a new one
2. If you don't find anything anywhere, submit an issue to the project repository explaining what you found and how you expected the program to behave and how the program behaves in your own environment
3. Before you submit the patch, you might want to mention the potential solution in the GitHub Issue. Sometimes I find it useful when I say something like this:
 ```
 I found that the problem is in this line () and that it would work if I changed it to this (), would you be interested in a pull request with this solution?
 ```
4. Submit a PR with a potential solution

> 5. Wait for feedback and make changes to your PR if the maintainers of the project have any until the PR is ready to merge!

Reviewer: ```suggestion
5. Wait for feedback. If the maintainers of the project have any suggestions make changes to your PR until the PR is ready to merge!
```

### If you think something would be a nice feature

1. Search in Google if the feaure was already proposed or what other people think of your feature; if the feature is already proposed and you think you can improve it, comment your ideas about it
2. If you don't find anything in the web, issue tracker, etc, submit an issue explaining why you think your feature will add value to the project

> 3. Wait for feedback from the project maintainers and learn about why your feature is or not addecuate to the project

Reviewer: ```suggestion
3. Wait for feedback from the project maintainers and learn about why your feature is or not adequate to the project
```

4. If you got positive feedback about your feature, start working on it. Don't be scared to ask for help if you get stuck. Finally submit a PR!
5. Wait for feedback again! You are almost there! Maintainers will give you more feedback about your feature; don't take it personal and respectfully discuss feedback. If you disagree with something it might be a good opportunity to learn from someone else's experience

### Triaging issue

> You might not have a feature idea or issue to report. That shouldn't stop you from contributing to an open source project. You could triage open issues and pull requests. That's another great way you could start contributing. Start checking if there are issues that are already solved and notify maintainers that the issue should be closed.  You could give feedback about a feature proposal. You could start reviewing existing PRs! The main goal here is to lose your fear and motivate yourself to get involved in your favorite project.

Reviewer: ```suggestion
You might not have a feature idea or issue to report. That shouldn't stop you from contributing to an open source project. You could triage open issues and pull requests. That's another great way you could start contributing. Start checking if there are issues that are already solved and notify maintainers that the issue should be closed. You could give feedback about a feature proposal. You could start reviewing existing PRs! The main goal here is to lose your fear and motivate yourself to get involved in your favorite project.
```
